base/strings/styled.jl

```diff
@@ -0,0 +1,388 @@
+
+StyledString(s::AbstractString, props::Vector{<:Tuple{UnitRange{Int}, <:Pair{Symbol, <:Any}}}) =
+    StyledString(s, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}(props))
+
+StyledChar(c::AbstractChar, prop::Pair{Symbol, <:Any}, props::Pair{Symbol, <:Any}...) =
+    StyledChar(c, Vector{Pair{Symbol, Any}}(vcat(prop, props...)))
+
+# Constructors to avoid recursive wrapping
+
+StyledString(s::StyledString, props::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}) =
+    StyledString(s.string, vcat(s.properties, props))
+
+StyledChar(c::StyledChar, props::Vector{Pair{Symbol, Any}}) =
+    StyledChar(c.char, vcat(s.properties, props))
+
+# To avoid pointless overhead
+String(s::StyledString{String}) = s.string
+
+## Conversion/promotion ##
+
+convert(::Type{StyledString}, s::StyledString) = s
+convert(::Type{StyledString{S}}, s::S) where {S <: AbstractString} =
+    StyledString(s, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+convert(::Type{StyledString}, s::S) where {S <: AbstractString} =
+    convert(StyledString{S}, s)
+StyledString(s::S) where {S <: AbstractString} = convert(StyledString{S}, s)
+
+convert(::Type{StyledChar}, c::StyledChar) = c
+convert(::Type{StyledChar{C}}, c::C) where { C <: AbstractChar } =
+    StyledChar{C}(c, Vector{Pair{Symbol, Any}}())
+convert(::Type{StyledChar}, c::C) where { C <: AbstractChar } =
+    convert(StyledChar{C}, c)
+
+StyledChar(c::AbstractChar) = convert(StyledChar, c)
+StyledChar(c::UInt32) = convert(StyledChar, Char(c))
+StyledChar{C}(c::UInt32) where {C <: AbstractChar} = convert(StyledChar, C(c))
+
+promote_rule(::Type{<:StyledString}, ::Type{<:AbstractString}) = StyledString
+
+## AbstractString interface ##
+
+ncodeunits(s::StyledString) = ncodeunits(s.string)
+codeunits(s::StyledString) = codeunits(s.string)
+codeunit(s::StyledString) = codeunit(s.string)
+codeunit(s::StyledString, i::Integer) = codeunit(s.string, i)
+isvalid(s::StyledString, i::Integer) = isvalid(s.string, i)
+@propagate_inbounds iterate(s::StyledString, i::Integer=firstindex(s)) =
+    if i <= lastindex(s.string); (s[i], nextind(s, i)) end
+eltype(::Type{<:StyledString{S}}) where {S} = StyledChar{eltype(S)}
+firstindex(s::StyledString) = firstindex(s.string)
+lastindex(s::StyledString) = lastindex(s.string)
+
+function getindex(s::StyledString, i::Integer)
+    @boundscheck checkbounds(s, i)
+    @inbounds if isvalid(s, i)
+        StyledChar(s.string[i], textproperties(s, i))
+    else
+        string_index_err(s, i)
+    end
+end
+
+## AbstractChar interface ##
+
+ncodeunits(c::StyledChar) = ncodeunits(c.char)
+codepoint(c::StyledChar) = codepoint(c.char)
+
+# Avoid the iteration fallback with comparison
+cmp(a::StyledString, b::AbstractString) = cmp(a.string, b)
+cmp(a::AbstractString, b::StyledString) = cmp(a, b.string)
+# To avoid method ambiguity
+cmp(a::StyledString, b::StyledString) = cmp(a.string, b.string)
+
+==(a::StyledString, b::StyledString) =
+    a.string == b.string && a.properties == b.properties
```

Comment:
Do we want this, instead of just comparing the underlying string? (I.e. I would almost always consider two strings equal if they have the same word-for-word content). And if we do want this, does the order of the elements in the properties matter?

Reply:
This is something I'm not sure, as it depends on the notion of equality used.

Reply:
One approach: First test the strings. If they are the same, then run `styledstring_optimize!` on both strings, then check if they have the same number of elements in the properties vector. Then  `issetequal` on the properties vector.

Reply:
Sure, but what I'm reconsidering is what `==` should be for `StyledStrings`. If we just compare the underlying string with `==`, what operator is appropriate when checking properties too?

Reply:
Hmm, would it be unreasonable to define this as `isequal`?

Reply:
> I would almost always consider two strings equal if they have the same word-for-word content

I wouldn't do that for the same reason I'd consider `OrderedDict`s with different orders unequal. I expect that `x::T, y::T, f`, for any function `f` in the API of `T`, we have `x == y implies f(x) == f(y)`. Since the style is in the API of `StyledString`, I expect two strings with different styles to be unequal.

Reply:
Ok, I'm currently leaving this as same-string-with-same-properties. Regarding ordering, most of the time the properties _should_ be in sorted order, which should make things easier. `::StyledString == ::AbstactString` now holds if the `StyledString` has no properties and `==` holds on the underlying string.